test/lock.cc

```diff
@@ -347,34 +292,103 @@ TEST(LockTest, ConditionVariable) {
 
   future2.get();
   future3.get();
 }
+
+TEST(LockTest, ConditionVariableUseAfterFree) {
+  // A bug was caught in the wild where `ConditionVariable` would enqueue a
+  // waiting `eventual` for later notification, even if the waiting condition
+  // was already met at the time of creation. This would result in a
+  // use-after-free type bug where the `ConditionVariable` at a later point
+  // attempts to notify a stale eventual.
+  //
+  // The bug is demonstrated below using a simple binary condition object
+  // `Foo`, that can be either in the state `true` or `false`. Starting with
+  // the condition variable in the false state, an `eventual` is created (and
+  // started) that will simply wait for the `false` state and return the
+  // current state. This is non blocking and should not result in the
+  // `eventual` getting enqueued in the `ConditionVariable` for later
+  // notification. The eventual is then garbage collected.
+  //
+  // The state of the condition object is then "updated" to `false` and all the
+  // internal `waiter`s are notified. In case the destructed eventual is in the
+  // list the test will blow up with a message along the lines of
+  // "pure virtual method called" as the `ConditionVariable` attempts to
+  // `notify` the destructed object.
+
+  struct Foo : public Synchronizable {
+    Foo() : condition_variable_(&lock()) {}
+
+    auto SetCondition(bool value) {
+      return Synchronized(Then([this, value]() mutable {
+        condition_ = value;
+        condition_variable_.NotifyAll();
+        return value;
+      }));
+    }
+
+    auto WaitFor(bool value) {
+      return Synchronized(condition_variable_.Wait([this, value]() {
+        return condition_ != value;
+      }) | Just(value));
+    }
+
+    bool condition_ = false;
+    ConditionVariable condition_variable_;
+  };
+
+  // Create instance of binary condition.
+  Foo foo;
+
+  {
+    // The anonymous scope is deliberate to ensure that the constructed
+    // `eventual` is destructed in a few lines.
+
+    // Wait for `false`. This should be a no-op and not result in the
+    // `eventual` being enqueued.
+    auto [future, k] = Terminate(foo.WaitFor(false));
+    k.Start();
+
+    EXPECT_NE(std::future_status::timeout,
+              future.wait_for(std::chrono::seconds(0)));
+
+    bool condition_value = future.get();
+    EXPECT_EQ(condition_value, false);
+  }
+
+  // In case the (now destructed) eventual is enqueued internally as a `waiter`
+  // in the condition variable, setting the condition variable to `false`
+  // (again) and attempting to notify the destructed object should make the
+  // test blow up.
+  auto [update_future, update_k] = Terminate(foo.SetCondition(false));
+  update_k.Start(); // In case of error, this blows up.
+
+  // Else, make sure we are not blocked, and terminate test.
+  EXPECT_NE(std::future_status::timeout,
+            update_future.wait_for(std::chrono::seconds(0)));
+  update_future.get();
```

Comment:
As commented above, since you're waiting here no matter what, how about just:
```suggestion
  *foo.SetCondition(false);
```